trivup/apps/SslApp.py

```diff
@@ -199,17 +220,76 @@ def create_cert(self, cn):
+
+        with open(ssl_cfg, 'w') as f:
+            f.write(dedent("""
+                [req]
+                distinguished_name=dn
+                [ dn ]
+                [ ext ]
+                basicConstraints=CA:TRUE,pathlen:0
+            """))
+
+        self.dbg('Generating key for %s intermediate: %s' %
+                 (cn, ret['intermediate_priv']['pem']))
+        self.exec_cmd('openssl genrsa -out "%s" 2048' %  # noqa: E501
+                    (ret['intermediate_priv']['pem']))
+        self.dbg('Generating request for %s: %s' % (cn, ret['req']))
+        self.exec_cmd('openssl req -config "%s" -extensions ext -key "%s" -new -out "%s" -subj "%s"' %  # noqa: E501
+                (ssl_cfg, ret['intermediate_priv']['pem'], ret['intermediate_req'], self.mksubj('%s-intermediate' % (cn))))
+
+        # Work out if "intermediate" cert should be self-signed or actually signed by the CA.
+        if with_ca:
+            self.dbg('Signing key for %s intermediate with CA cert' % (cn))
+            self.exec_cmd('openssl x509 -req -extfile "%s" -extensions ext -passin "pass:%s" -in "%s" -CA "%s" -CAkey "%s" -CAserial "%s" -out "%s"' %  # noqa: E501
+                        (ssl_cfg, password,
+                        ret['intermediate_req'], self.ca['pem'], self.ca['key'],
+                        self.ca['srl'], ret['intermediate_pub']['pem']))
+        else:
+            self.dbg('Signing key for %s intermediate with self' % (cn))
+            self.exec_cmd('openssl x509 -req -passin "pass:%s" -in "%s" -signkey "%s" -out "%s"' %  # noqa: E501
+                        (password,
+                        ret['intermediate_req'], ret['intermediate_priv']['pem'], ret['intermediate_pub']['pem']))
+
+        self.dbg('Converting public-key X.509 to DER for %s intermediate' % cn)
+        self.exec_cmd('openssl x509 -outform der -in "%s" -out "%s"' %  # noqa: E501
+                    (ret['intermediate_pub']['pem'], ret['intermediate_pub']['der']))
+
+        self.dbg('Converting private-key X.509 to DER for %s intermediate' % cn)
+        self.exec_cmd('openssl rsa -outform der -passin "pass:%s" -in "%s" -out "%s"' %  # noqa: E501
+                    (password, ret['intermediate_priv']['pem'], ret['intermediate_priv']['der']))
+
+    def _export_pkcs12(self, ret, cn, through_intermediate, with_ca):
+        password = self.conf.get('ssl_key_pass')
+        additional_certs_for_pkcs12 = []
+        if through_intermediate:
+            additional_certs_for_pkcs12.append(ret['intermediate_pub']['pem'])
+        if with_ca:
+            additional_certs_for_pkcs12.append(self.ca['pem'])
+        certfile_arguments = ' '.join(
+            ['-certfile "{}"'.format(c) for c in additional_certs_for_pkcs12])
+
         self.dbg('Creating PKCS#12 for %s in %s' % (cn, ret['pkcs']))
-        self.exec_cmd('openssl pkcs12 -export -out "%s" -inkey "%s" -in "%s" -CAfile "%s" -certfile "%s" -passin "pass:%s" -passout "pass:%s"' %  # noqa: E501
```

Comment:
Is `-certfile ca.pem` identical to `-CAfile ca.pem` ?

Reply:
I believe, on my reading of the `openssl pkcs12` man page, that our usage of `-CAfile` here actually does nothing. I'm pretty sure that the purpose of the `-CAfile` and `-CApath` arguments is so that the `-chain` argument can extract relevant CA certificates out of the CA store and put them in the exported pkcs12 file (see for example https://arstechnica.com/civis/viewtopic.php?p=24680099). However, we're not passing `-chain` here, and we're unconditionally including the CA certificates in the pkcs12 file with `-certfile` anyway, so I think the `-CAfile` argument here was serving no purpose.

Reply:
Thanks for sorting that out :+1: